File: herminedb-driver/src/main/java/org/hermine/internal/db/driver/SequentialOperationGroup.java
/*
 * Copyright 2018 HermineDB's author : Frédéric Montariol,
 * and explicitly declared author of this file if provided.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.hermine.internal.db.driver;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Scheduler;

class SequentialOperationGroup<S, T> extends AbstractOperationGroup<S, T> {

    private static final Logger LOGGER = LogManager.getLogger(SequentialOperationGroup.class);

    boolean heldFinished = false;

    /**
     * completed when this OperationGroup is no longer held. Completion of this
     * OperationGroup depends on held.
     *
     * @see #submit,
     * @see #releaseProhibitingMoreMembers,
     * @see #submitHoldingForMoreMembers
     */
    private Mono held;

    /**
     * predecessor of all member Operations and the OperationGroup itself
     */
    private Mono head;

    /**
     * The last Mono of any submitted member Operation. Mutable until
     * not isHeld().
     */
    private Mono<S> memberTail;

    SequentialOperationGroup(ConnectionImpl conn, AbstractOperationGroup<? super T, ?> group) {
        super(conn, group);
        held = Mono.empty()
            .doOnSuccessOrError((o, throwable) -> heldFinished = true);
        head = Mono.empty();
        memberTail = head;
    }

    // Override OperationGroup methods

    @Override
    public SubmissionImpl<T> releaseProhibitingMoreMembers() {
        LOGGER.info("SequentialOperationGroup#releaseProhibitingMoreMembers");
        if ( ! isImmutable() || ! isHeld() ) throw new IllegalStateException("TODO");
        held.subscribe();
        immutable();  // having set isHeld to false this call will make this OpGrp immutable
        return submission;
    }

    @Override
    public SubmissionImpl<T> submit() {
        LOGGER.info("SequentialOperationGroup#submit");
        if ( isImmutable() ) throw new IllegalStateException("TODO");
        accumulator = collector.supplier().get();
        SubmissionImpl<T> sub = super.submit();
        held.subscribe();
        return sub;
    }

    // Internal methods

    @Override
    Mono<T> follows(Mono<?> predecessor, Scheduler scheduler) {
        LOGGER.info("SequentialOperationGroup#follows");
        return condition.flatMap(cond -> {
                    if (cond) {
                        head = head.thenReturn(predecessor);
                        held = held.then(memberTail.map(t -> (T)collector.finisher()
                                        .apply(accumulator))
                                .publishOn(scheduler)
                        );
                        return held;
                    }
                    else {
                        return Mono.empty();
                    }
                }
        );
    }

    @Override
    protected boolean isHeld() {
        LOGGER.info("SequentialOperationGroup#isHeld {}", !heldFinished);
        return !heldFinished;
    }
}

File: herminedb-driver/src/main/kotlin/org/hermine/internal/db/driver/SequentialOperationGroup.kt
/*
 * Copyright 2018 HermineDB's author : Frédéric Montariol
 * and explicitly declared author of each file if specified.
 * Use of this source code is governed by the Apache 2.0 license.
 */

package org.hermine.internal.db.driver

import kotlinx.coroutines.experimental.CompletableDeferred
import kotlinx.coroutines.experimental.Deferred
import kotlinx.coroutines.experimental.async
import mu.KotlinLogging
import java.time.Duration
import java.util.function.Consumer
import java.util.function.Function
import kotlin.coroutines.experimental.CoroutineContext

// Place definition above class declaration to make field static
private val logger = KotlinLogging.logger {}

internal open class SequentialOperationGroup<S, T> : AbstractOperationGroup<S, T> {

    // used only by Session. Will break if used by any other class.
    constructor() : super()

    constructor(conn: ConnectionImpl, group: AbstractOperationGroup<in T, *>) : super(conn, group)

    /**
     * completed when this OperationGroup is no longer held. Completion of this
     * OperationGroup depends on held.
     *
     * @see #submit,
     * @see #releaseProhibitingMoreMembers,
     * @see #submitHoldingForMoreMembers
     */
    private val held: CompletableDeferred<Unit> = CompletableDeferred()

    override fun releaseProhibitingMoreMembers(): SubmissionImpl<T> {
        logger.debug{"SequentialOperationGroup#releaseProhibitingMoreMembers"}
        if (!isImmutable() || !isHeld()) throw IllegalStateException("TODO")
        held.complete(Unit)
        immutable()  // having set isHeld to false this call will make this OpGrp immutable
        return submission!!
    }

    // Override Operation methods

    override fun timeout(minTime: Duration): SequentialOperationGroup<S, T> {
        super.timeout(minTime)
        return this
    }

    override fun onError(handler: Consumer<Throwable>): SequentialOperationGroup<S, T> {
        super.onError(handler)
        return this
    }

    override fun submit(): SubmissionImpl<T> {
        logger.debug{"SequentialOperationGroup#submit"}
        if (isImmutable()) throw IllegalStateException("TODO")
        accumulator = collector.supplier().get()
        held.complete(Unit)
        return super.submit()
    }

    // Internal methods

    override fun follows(predecessor: Deferred<*>?, context: CoroutineContext): Deferred<T?> {
        logger.debug{"SequentialOperationGroup#follows"}
        return async(context) {
            val cond = condition.await()
            if (cond) {
                head.complete(predecessor)
                held.await()
                (collector.finisher() as Function<Any?, Any>)
                        .apply(accumulator) as T
            } else {
                null
            }
        }
    }

    override fun isHeld() = !held.isCompleted
}
